### PR TITLE
hal/egl fix support for context extensions

### LINKED ISSUE #1577: [GL] Unable to create GLES 3.x context on Android
**Description**
Unable to create GLES 3.x context on Android device.

```
2021-06-30 22:17:11.584 16981-16981/rust.example.cube V/ndk_glue: WindowCreated
2021-06-30 22:17:11.584 16981-17057/rust.example.cube I/cube::framework: NativeScreen Found:Some(NativeWindow { ptr: 0x76f6467010 })
2021-06-30 22:17:11.584 16981-17057/rust.example.cube I/cube::framework: Initializing the surface...
2021-06-30 22:17:11.586 16981-16981/rust.example.cube V/ndk_glue: WindowResized
2021-06-30 22:17:11.594 16981-17057/rust.example.cube D/wgpu_hal::gles::egl: Client extensions: []
2021-06-30 22:17:11.594 16981-17057/rust.example.cube I/wgpu_hal::gles::egl: Using default platform
2021-06-30 22:17:11.596 16981-16981/rust.example.cube V/ndk_glue: ContentRectChanged
2021-06-30 22:17:11.615 16981-17057/rust.example.cube I/wgpu_hal::gles::egl: Display vendor "Android", version (1, 4)
2021-06-30 22:17:11.615 16981-17057/rust.example.cube D/wgpu_hal::gles::egl: Display extensions: [
        "EGL_KHR_get_all_proc_addresses",
        "EGL_ANDROID_presentation_time",
        "EGL_KHR_swap_buffers_with_damage",
        "EGL_ANDROID_get_native_client_buffer",
        "EGL_ANDROID_front_buffer_auto_refresh",
        "EGL_ANDROID_get_frame_timestamps",
        "EGL_KHR_image",
        "EGL_KHR_image_base",
        "EGL_KHR_gl_colorspace",
        "EGL_KHR_gl_texture_2D_image",
        "EGL_KHR_gl_texture_cubemap_image",
        "EGL_KHR_gl_renderbuffer_image",
        "EGL_KHR_fence_sync",
        "EGL_KHR_create_context",
        "EGL_KHR_config_attribs",
        "EGL_KHR_surfaceless_context",
        "EGL_EXT_create_context_robustness",
        "EGL_ANDROID_image_native_buffer",
        "EGL_KHR_wait_sync",
        "EGL_ANDROID_recordable",
        "EGL_KHR_partial_update",
        "EGL_KHR_mutable_render_buffer",
        "EGL_EXT_protected_content",
        "EGL_IMG_context_priority",
    ]
2021-06-30 22:17:11.616 16981-17057/rust.example.cube V/wgpu_hal::gles::egl: Configurations:
2021-06-30 22:17:11.616 16981-17057/rust.example.cube V/wgpu_hal::gles::egl: 	CONFORMANT=0x45, RENDERABLE=0x45, NATIVE_RENDERABLE=0x0, SURFACE_TYPE=0x1405
2021-06-30 22:17:11.616 16981-17057/rust.example.cube I/chatty: uid=10225(u0_a225) rust.example.cube identical 1 line
2021-06-30 22:17:11.616 16981-17057/rust.example.cube V/wgpu_hal::gles::egl: 	CONFORMANT=0x45, RENDERABLE=0x45, NATIVE_RENDERABLE=0x0, SURFACE_TYPE=0x1405
2021-06-30 22:17:11.616 16981-17057/rust.example.cube V/wgpu_hal::gles::egl: 	CONFORMANT=0x45, RENDERABLE=0x45, NATIVE_RENDERABLE=0x0, SURFACE_TYPE=0x405
2021-06-30 22:17:11.616 16981-17057/rust.example.cube V/wgpu_hal::gles::egl: 	CONFORMANT=0x45, RENDERABLE=0x45, NATIVE_RENDERABLE=0x0, SURFACE_TYPE=0x1405
2021-06-30 22:17:11.616 16981-17057/rust.example.cube I/chatty: uid=10225(u0_a225) rust.example.cube identical 1 line
2021-06-30 22:17:11.616 16981-17057/rust.example.cube V/wgpu_hal::gles::egl: 	CONFORMANT=0x45, RENDERABLE=0x45, NATIVE_RENDERABLE=0x0, SURFACE_TYPE=0x1405
2021-06-30 22:17:11.616 16981-17057/rust.example.cube V/wgpu_hal::gles::egl: 	CONFORMANT=0x45, RENDERABLE=0x45, NATIVE_RENDERABLE=0x0, SURFACE_TYPE=0x405
2021-06-30 22:17:11.616 16981-17057/rust.example.cube V/wgpu_hal::gles::egl: 	CONFORMANT=0x45, RENDERABLE=0x45, NATIVE_RENDERABLE=0x0, SURFACE_TYPE=0x1405
2021-06-30 22:17:11.616 16981-17057/rust.example.cube V/wgpu_hal::gles::egl: 	CONFORMANT=0x45, RENDERABLE=0x45, NATIVE_RENDERABLE=0x0, SURFACE_TYPE=0x1405
2021-06-30 22:17:11.616 16981-17057/rust.example.cube V/wgpu_hal::gles::egl: 	CONFORMANT=0x45, RENDERABLE=0x45, NATIVE_RENDERABLE=0x0, SURFACE_TYPE=0x405
2021-06-30 22:17:11.616 16981-17057/rust.example.cube V/wgpu_hal::gles::egl: 	CONFORMANT=0x45, RENDERABLE=0x45, NATIVE_RENDERABLE=0x0, SURFACE_TYPE=0x1
2021-06-30 22:17:11.616 16981-17057/rust.example.cube I/chatty: uid=10225(u0_a225) rust.example.cube identical 2 lines
2021-06-30 22:17:11.616 16981-17057/rust.example.cube V/wgpu_hal::gles::egl: 	CONFORMANT=0x45, RENDERABLE=0x45, NATIVE_RENDERABLE=0x0, SURFACE_TYPE=0x1
2021-06-30 22:17:11.616 16981-17057/rust.example.cube I/wgpu_hal::gles::egl: 	Trying presentation
2021-06-30 22:17:11.617 16981-17057/rust.example.cube I/wgpu_hal::gles::egl: 	EGL context: +robust access
2021-06-30 22:17:11.617 16981-17057/rust.example.cube W/wgpu_hal::gles::egl: unable to create GLES 3.x context: BadAttribute
2021-06-30 22:17:11.619 16981-17056/rust.example.cube I/RustStdoutStderr: thread '<unnamed>' panicked at 'No suitable GPU adapters found on the system!', wgpu/examples/cube/../framework.rs:138:10
```

**Platform**
Android 8.0.

### PR DESCRIPTION
**Pull request by kvark**
**Connections**
Follow-up to #1584
Closes  #1577

**Description**
TIL that EGL extensions work differently from the EGL-1.5 stuff, and we are now treating them properly.

**Testing**
Basic testing on AMD

### CODE REVIEW

File: wgpu-hal/src/gles/egl.rs [hunk 1]
```diff
@@ -260,24 +264,41 @@ impl Inner {
...
+            if version >= (1, 5) {
+                context_attributes.push(egl::CONTEXT_OPENGL_DEBUG);
+                context_attributes.push(egl::TRUE as _);
+            } else if supports_khr_context {
+                khr_context_flags |= EGL_CONTEXT_OPENGL_DEBUG_BIT_KHR;
+            }
         }
-        if display_extensions.contains("EGL_EXT_create_context_robustness") {
+        if needs_robustness {
             log::info!("\tEGL context: +robust access");
```
Comment (cwfitzgerald) [inline wgpu-hal/src/gles/egl.rs:286]: could these debugs be moved into the if statements and carry some info about which version they're enabling, like for robustness they could say `+robust access`, `+robust access EXT`, and `+robust access KHR`.

File: wgpu-hal/src/gles/egl.rs [hunk 2]
```diff
@@ -260,24 +264,48 @@ impl Inner {
...
+            if version >= (1, 5) {
+                log::info!("\tEGL context: +robust access");
+                context_attributes.push(egl::CONTEXT_OPENGL_ROBUST_ACCESS);
+                context_attributes.push(egl::TRUE as _);
+            } else if supports_khr_context {
+                log::info!("\tEGL context: +robust access KHR");
+                khr_context_flags |= EGL_CONTEXT_OPENGL_ROBUST_ACCESS_BIT_KHR;
+            } else if display_extensions.contains("EGL_EXT_create_context_robustness") {
+                log::info!("\tEGL context: +robust access EXT");
+                context_attributes.push(EGL_CONTEXT_OPENGL_ROBUST_ACCESS_EXT);
```
Comment (Gordon-F) [inline wgpu-hal/src/gles/egl.rs:298]: @kvark Thank you! But this not  fix #1577 on my Android device (8.0, EGL 1.4). Still got same `BadAttibute` error.

**Review by Gordon-F — APPROVED** [review]

File: wgpu-hal/src/gles/egl.rs (hunk 2 above)
Comment (kvark) [reply wgpu-hal/src/gles/egl.rs:298]: @Gordon-F could you attach a new log to the issue with this PR? Curious about what path is taken.

**Comment by cwfitzgerald** [comment]
Works on my machine!

bors r+

**Pull request merged**